message/router.go

```diff
@@ -216,6 +239,27 @@ func (r *Router) AddHandler(
 	}
 }
 
+// AddMiddlewareToHandler adds a new middleware to specified handler in the router.
+//
+// The order of middlewares matters. Middleware added at the beginning is executed first.
+func (r *Router) AddMiddlewareToHandler(handlerName string, m ...HandlerMiddleware) {
```

Comment:
Can you consider adding some high-level interface for this? I feel like basing this on handler name can be fragile and not that convenient.

Something similar to the go-chi way would be:

```go
router.Route(func (r Router) {
    r.AddMiddleware(middleware.Retry)
    r.AddHandler(myHandler)
})
```

But that can be quite hard to implement and maybe overcomplicated.

Another idea might be an interface similar to structured loggers:

```go
router.With(middleware.Retry).AddHandler(handler)

// Maybe even with variadic arguments
router.With(
    middleware.Recoverer,
    middleware.Retry,
).AddHandler(myHandler)
```

Any thoughts? :)

Reply:
Thank you for the clarification.

In my opinion, the first approach might be a bit misleading since, in fact, there is no route (subpath) here. I personally prefer the second approach.

Could you tell me what is the expected flow of execution?
```
...
router.AddMiddleware(middlewareA)
router.AddHandler("handler_1",...)
router.AddMiddleware(middlewareB)
...
```
Currently (without middleware per handler) the flow looks like this:
`middlewareA => middlewareB => handler_1`

__What is the expected flow of the execution in such case?__
```
...
router.AddMiddleware(middlewareA)
router.With(
    middlewareC,
    middlewareD,
).AddHandler("handler_1",...)
router.AddMiddleware(middlewareB)
...
```

__Should it be:__
* `middlewareA => middlewareB => middlewareC => middlewareD => handler_1` or
* `middlewareA => middlewareC => middlewareD => middlewareB => handler_1` or
* `middlewareA => middlewareC => middlewareD => handler_1` (this one contradicts with the current solution, but it is similar to the approach used in go-chi where adding middleware after handler is forbidden https://github.com/go-chi/chi/blob/2171e2a8965db606ce84d224841987315c0eb515/mux.go#L93)

Reply:
That's interesting question and I'm not sure what's the most obvious for end users. I feel like the way chi handles it (with an error) seems the most idiomatic.

@roblaszczak @maclav3 any thoughts from you guys?

Reply:
```
router.Route(func (r Router) {
    r.AddMiddleware(middleware.Retry)
    r.AddHandler(myHandler)
})
```

```router.With(middleware.Retry).AddHandler(handler)

// Maybe even with variadic arguments
router.With(
    middleware.Recoverer,
    middleware.Retry,
).AddHandler(myHandler)
```
this looks better, but IMO a bit wired (it can be confused with AddMiddleware)


and about order my guess is `middlewareA > MiddlewareC > MiddlewareD > Middleware B` (so we should probably have some global middleware counter?)


yup, it looks overcomplicated (in chi it make sense because you have sub-routers)

and different idea is, maybe we can return some builder with handler, so it may be

```
handler1 := router.AddHandler("handler_1", ...)
handler1.AddMiddleware(middlewareA)
```

Reply:
so for the third option, I can imagine that AddHandler may return some kind of builder or `handler` object (with needs to be still connected to the router - to for example keep middlewares order) and may store internally middlewares added to handler

Reply:
@roblaszczak Thanks for the suggestions.